drivers/rtc/rtc_ds3231.c

```diff
@@ -0,0 +1,554 @@
+	timeptr->tm_min =
+		bcd2bin(regs[1] & DS3231_MINUTES_MASK) + bcd2bin(regs[1] & DS3231_MINUTES_10);
+	timeptr->tm_hour =
+		bcd2bin(regs[2] & DS3231_HOURS_MASK) + bcd2bin(regs[2] & DS3231_HOURS_10);
+	timeptr->tm_wday = bcd2bin(regs[3] & DS3231_DAYS_MASK);
+	timeptr->tm_mday = bcd2bin(regs[4] & DS3231_DATE_MASK) + bcd2bin(regs[4] & DS3231_DATE_10);
+	timeptr->tm_mon = bcd2bin(regs[5] & DS3231_MONTHS_MASK) +
+			  bcd2bin(regs[5] & DS3231_MONTH_10) - DS3231_MONTHS_OFFSET;
+	timeptr->tm_year = bcd2bin(regs[6] & DS3231_YEARS_MASK) +
+			   bcd2bin(regs[6] & DS3231_YEAR_10) + DS3231_YEARS_OFFSET;
+	LOG_DBG("get time: year = %d, mon = %d, mday = %d, wday = %d, hour = %d, "
+		"min = %d, sec = %d",
+		timeptr->tm_year, timeptr->tm_mon, timeptr->tm_mday, timeptr->tm_wday,
+		timeptr->tm_hour, timeptr->tm_min, timeptr->tm_sec);
+
+	return 0;
+}
+#ifdef CONFIG_RTC_ALARM
+static int ds3231_alarm_get_supported_fields(const struct device *dev, uint16_t id, uint16_t *mask)
+{
+
+	if (id == 0U) {
+		*mask = DS3231_RTC_ALARM_1_TIME_MASK;
+	} else if (id == 1U) {
+		*mask = DS3231_RTC_ALARM_2_TIME_MASK;
+	} else {
+		LOG_ERR("invalid ID %d", id);
+		return -EINVAL;
+	}
+	LOG_INF("Supported mask is " PRINTF_BINARY_PATTERN_INT16,
+		PRINTF_BYTE_TO_BINARY_INT16(*mask));
+	return 0;
+}
+
+static int ds3231_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+				 struct rtc_time *timeptr)
+{
+	int err;
+	uint8_t regs_0[4];
+	uint8_t regs_1[3];
+	if (id == 0u) {
+		err = ds3231_read_regs(dev, DS3231_ALARM_1_SECONDS, &regs_0, sizeof(regs_0));
+		if (err != 0) {
+			return err;
+		}
+		memset(timeptr, 0U, sizeof(*timeptr));
+		timeptr->tm_sec = bcd2bin(regs_0[0] & DS3231_ALARM_1_SECONDS_SECONDS) +
+				  bcd2bin(regs_0[0] & DS3231_ALARM_1_SECONDS_10);
+		timeptr->tm_min = bcd2bin(regs_0[1] & DS3231_ALARM_1_MINUTES_MINUTES) +
+				  bcd2bin(regs_0[1] & DS3231_ALARM_1_MINUTES_10);
+		timeptr->tm_hour = bcd2bin(regs_0[2] & DS3231_ALARM_1_HOURS_HOURS) +
+				   bcd2bin(regs_0[2] & DS3231_ALARM_1_HOURS_10) +
+				   bcd2bin(regs_0[2] & DS3231_ALARM_1_HOURS_AM_PM_20);
+		timeptr->tm_mday = bcd2bin(regs_0[3] & DS3231_ALARM_1_DAY_DATE_MASK) +
+				   bcd2bin(regs_0[3] & DS3231_ALARM_1_DAY_DATE_10);
+	} // endif id=0
+	else if (id==1U){
+		err = ds3231_read_regs(dev, DS3231_ALARM_2_MINUTES, &regs_1, sizeof(regs_1));
+		if (err != 0) {
+			return err;
+		}
+		memset(timeptr, 0U, sizeof(*timeptr));
+		timeptr->tm_min = bcd2bin(regs_1[0] & DS3231_ALARM_2_MINUTES_MINUTES) +
+				  bcd2bin(regs_1[0] & DS3231_ALARM_2_MINUTES_10);
+		timeptr->tm_hour = bcd2bin(regs_1[1] & DS3231_ALARM_2_HOURS_HOURS) +
+				   bcd2bin(regs_1[1] & DS3231_ALARM_2_HOURS_10) +
+				   bcd2bin(regs_1[1] & DS3231_ALARM_2_HOURS_AM_PM_20);
+		timeptr->tm_mday = bcd2bin(regs_1[2] & DS3231_ALARM_2_DAY_DATE_MASK) +
+				   bcd2bin(regs_1[2] & DS3231_ALARM_2_DAY_DATE_10);
+
+	}
+	return 0;
+}
+
+static int ds3231_alarm_is_pending(const struct device *dev, uint16_t id)
+{
+	return 0;
```

Comment:
I'm aware you probably have this staged to add but just a reminder that it's missing